gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireContentsProcedure.java

```diff
@@ -0,0 +1,280 @@
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow internalRow) {
+    String gcCatalogName = internalRow.getString(0);
+    String gcOutputBranchName = internalRow.getString(1);
+    String gcOutputTableIdentifier = internalRow.getString(2);
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(3);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(4) ? internalRow.getString(4) : null;
+    boolean dryRun = !internalRow.isNullAt(5) && internalRow.getBoolean(5);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    if (runId == null) {
+      runId = getLatestCompletedRunID(gcOutputTableIdentifier, identifiedResultsRepo);
+    }
+
+    FileIO fileIO = getFileIO(nessieClientConfig);
+    Dataset<Row> expiredContents = getExpiredContents(runId, identifiedResultsRepo, fileIO);
+    if (!dryRun) {
+      expiredContents =
+          expiredContents.map(
+              new DeleteFunction(fileIO), RowEncoder.apply(expiredContents.schema()));
+    }
+
+    List<Row> rows = expiredContents.collectAsList();
+
+    List<InternalRow> outputRows = new ArrayList<>();
+    rows.forEach(
+        row ->
+            outputRows.add(
+                GCProcedureUtil.internalRow(
+                    // "content_id", "deleted_files_type", "deleted_files_count",
+                    // "deleted_files_list"
+                    row.getString(0), row.getString(1), row.getInt(3), row.getList(2))));
+
+    return outputRows.toArray(new InternalRow[0]);
+  }
+
+  private FileIO getFileIO(Map<String, String> nessieClientConfig) {
+    Configuration config = spark().sparkContext().hadoopConfiguration();
+    String fileIOImpl = nessieClientConfig.get(CatalogProperties.FILE_IO_IMPL);
+    return fileIOImpl == null
+        ? new HadoopFileIO(config)
+        : CatalogUtil.loadFileIO(fileIOImpl, nessieClientConfig, config);
+  }
+
+  private static Dataset<Row> getExpiredContents(
+      String runId, IdentifiedResultsRepo identifiedResultsRepo, FileIO fileIO) {
+    // Read the expired content-rows from output table for this run id
+    Dataset<Row> expiredContents = identifiedResultsRepo.collectExpiredContentsAsDataSet(runId);
+    Dataset<Row> expiredContentsDF = computeAllFiles(fileIO, expiredContents);
+    // Read the live content-rows from output table for this run id
+    Dataset<Row> liveContents = identifiedResultsRepo.collectLiveContentsAsDataSet(runId);
+    Dataset<Row> liveContentsDF = computeAllFiles(fileIO, liveContents);
+    // remove the files which are used by live contents
+    Dataset<Row> expiredFilesDF = expiredContentsDF.except(liveContentsDF);
```

Comment:
Does this really ensure that a data file that is referenced in an expired snapshot's manifest _and_ in a live snapshot's manifest is _not_ deleted?

rewrite_manifests?